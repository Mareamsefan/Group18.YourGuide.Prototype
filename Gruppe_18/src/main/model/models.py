from sqlalchemy import Table, Column, String, Integer, ForeignKey, DATETIME
from sqlalchemy.orm import relationship
from flask_login import UserMixin
import uuid

from Gruppe_18.src.main.database.sql_alchemy import db

tour_account_association = Table(
    'tour_account_association', db.metadata,
    Column('tour_id', String, ForeignKey('Tour.id')),
    Column('account_id', String, ForeignKey('User.id'))
)


class Account(db.Model, UserMixin):
    __tablename__ = "User"
    id = db.Column(db.String, primary_key=True)
    usertype = Column("usertype", String)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)
    tours = relationship("Tour", secondary=tour_account_association, back_populates="participants")

    def get_id(self):
        return str(self.id)

    def __init__(self, usertype, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.usertype = usertype
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"


class Tour(db.Model):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)
    participants = relationship("Account", secondary=tour_account_association, back_populates="tours")

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

    def get_id(self):
        return str(self.id)
