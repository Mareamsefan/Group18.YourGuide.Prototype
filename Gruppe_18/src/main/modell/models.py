from sqlalchemy import Column, String, Integer, DATETIME
import uuid
from Gruppe_18.src.main.database.sql_alchemy import Base


class Account(Base):
    __tablename__ = "User"
    account_id = Column("id", String, primary_key=True)
    username = Column("username", String)
    password = Column("password", String)
    phoneNumber = Column("phoneNumber", String)
    emailAddress = Column("emailAddress", String)

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.account_id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress

    def __repr__(self):
        return f"({self.account_id}) {self.username} {self.password} {self.phoneNumber} {self.emailAddress}"

    def get_id(self):
        return self.id


class Tour(Base):
    # duration in hours
    # cost in dollars
    __tablename__ = "Tour"
    tour_id = Column("id", String, primary_key=True)
    title = Column("title", String)
    date = Column("date", DATETIME)
    destination = Column("destination", String)
    duration = Column("duration", Integer)
    cost = Column("cost", Integer)
    max_travelers = Column("max_travelers", Integer)
    language = Column("language", String)
    pictureURL = Column("pictureURL", String)
    booked = Column("booked", Integer, default=0)

    def __init__(self, title, date, destination, duration, cost, max_travelers, language, pictureURL):
        self.tour_id = str(uuid.uuid4())
        self.title = title
        self.date = date
        self.destination = destination
        self.duration = duration
        self.cost = cost
        self.language = language
        self.max_travelers = max_travelers
        self.pictureURL = pictureURL

    def __repr__(self):
        return f"({self.tour_id}) {self.title} {self.destination} {self.duration} {self.cost} {self.language} {self.max_travelers} {self.pictureURL} {self.booked}"

