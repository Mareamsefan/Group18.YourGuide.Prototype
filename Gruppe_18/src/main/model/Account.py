import json
import uuid


class Account:

    def __init__(self, username, password, phoneNumber, emailAddress):
        self.id = str(uuid.uuid4())
        self.username = username
        self.password = password
        self.phoneNumber = phoneNumber
        self.emailAddress = emailAddress



    def get_id(self):
        return self.id
