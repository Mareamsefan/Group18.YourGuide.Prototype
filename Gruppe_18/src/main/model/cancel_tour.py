import datetime

class personTour:
    def __init__(self, tour_name, user_name, reference_number):
        self.tour_name = tour_name
        self.user_name = user_name
        self.reference_number = reference_number
        self.sign_up_time = datetime.datetime.now()
        self.is_cancelled = False

class TourAdmin:
    def __init__(self):
        self.tours = []


    def add_tour(self, tour_name, user_name, reference_number):
        tour = personTour(tour_name, user_name, reference_number)  # Use personTour instead of Tour
        self.tours.append(tour)
        print(f"User '{user_name}' signed up for '{tour_name}' tour with the reference number '{reference_number}'.")
        return True

    def cancel_tour(self, tour_name, user_name, reference_number):
        current_time = datetime.datetime.now()
        for tour in self.tours:
            if tour.tour_name == tour_name and tour.user_name == user_name and not tour.is_cancelled:
                time_difference = current_time - tour.sign_up_time
                if time_difference.total_seconds() <= 24 * 3600:  # 24 hours in seconds
                    return True
                    print(f"User '{user_name}' has canceled the '{tour_name}' tour with the reference number '{reference_number}' within 24 hours. No penalty.")
                else:
                    print(f"User '{user_name}' has canceled the '{tour_name}' tour with the reference number '{reference_number}' after 24 hours. Penalty may apply.")
                    return False
                tour.is_cancelled = True
                self.notify_guide_and_admin(tour)
                self.remove_tour_from_user(tour)
                self.send_cancellation_email(tour)
                return
        print(f"User '{user_name}' is not signed up for '{tour_name}' tour with the reference number '{reference_number}' or has already canceled.")
        return True

    def notify_guide_and_admin(self, tour):
        # Implement notification to tour guide and admin here
        print(f"Notifying guide and admin about the cancellation of '{tour.tour_name}' by '{tour.user_name}' with the reference number '{tour.reference_number}'.")

    def remove_tour_from_user(self, tour):
        # Implement removing the tour from the user's list here
        print(f"Removing '{tour.tour_name}' from '{tour.user_name}' 's tour list with the reference number '{tour.reference_number}'.")

    def send_cancellation_email(self, tour):
        # Implement sending an email about the cancellation here
        print(f"Sending cancellation email to '{tour.user_name}' for '{tour.tour_name}'.")

# Example usage:
tour_admin = TourAdmin()

# User signs up for a tour
tour_admin.add_tour("Tour 1", "Nicole", "DORMMB_18")

# User cancels a tour within 24 hours
tour_admin.cancel_tour("Tour 1", "Nicole", "DORMMB_18")

# User cancels a tour after 24 hours
# tour_admin.cancel_tour("Tour 1", "Nicole", "DORMMB_18")

