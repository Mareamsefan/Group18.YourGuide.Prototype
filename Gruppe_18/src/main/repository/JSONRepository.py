import json


class JSONRepository:

    def save_to_json(self, entity, filename):
        try:
            with open(filename, "r") as json_file:
                filedata = json.load(json_file)
        except FileNotFoundError:
            filedata = []

        duplicate_found = False
        for item in filedata:
            if item == entity.to_dict():
                duplicate_found = True
                break

        if not duplicate_found:
            data = entity.to_dict()
            filedata.append(data)
        with open(filename, "w") as json_file:
            json.dump(filedata, json_file, indent=4)

    def save_to_stream(self, io_stream):
        try:
            filedata = json.load(io_stream)
        except json.JSONDecodeError:
            filedata = []

        duplicate_found = False
        for item in filedata:
            if item == self.to_dict():
                duplicate_found = True
                break

        if not duplicate_found:
            data = self.to_dict()
            filedata.append(data)

        io_stream.truncate(0)
        io_stream.seek(0)
        json.dump(filedata, io_stream, indent=4)

    def to_dict(self, entity):
        entity_dict = {}
        for key, value in entity.__dict__.items():
            entity_dict[key] = value
        return entity_dict

    def read_from_stream(self, stream):
        items = json.load(stream)
        stream.seek(0)
        return items




    # def from_json(self, entity, ):
