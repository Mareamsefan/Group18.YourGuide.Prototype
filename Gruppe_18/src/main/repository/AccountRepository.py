from Gruppe_18.src.main.repository.JSONRepository import JSONRepository
from Gruppe_18.src.main.model.models import Account


class AccountRepository (JSONRepository):
    def delete_account(self, session, entity):
        account_to_delete = session.query(Account).filter_by(username=entity.username).first()

        if account_to_delete:
            session.delete(account_to_delete)
            session.commit()
            return True

        return False

    '''

    def delete_account(self, entity, io_stream):
        try:
                io_stream.seek(0)
                filedata = json.load(io_stream)
        except json.JSONDecodeError:
            filedata = []

        index_to_delete = None
        for i, item in enumerate(filedata):
            if item["username"] == entity.username:
                index_to_delete = i
                break

        if index_to_delete is not None:
            del filedata[index_to_delete]

            io_stream.seek(0)
            io_stream.truncate()
            json.dump(filedata, io_stream, indent=4)
            return True

        return False
'''
    # needs to be updated after database is implemented

    def successful_registration(self, entity, io_stream):
        self.save_to_stream(entity, io_stream)
        return True

