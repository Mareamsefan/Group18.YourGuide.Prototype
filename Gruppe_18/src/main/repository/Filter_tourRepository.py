from Gruppe_18.src.main.repository.JSONRepository import JSONRepository

class FilterRepository (JSONRepository):
    def read_from_stream(self, stream):
