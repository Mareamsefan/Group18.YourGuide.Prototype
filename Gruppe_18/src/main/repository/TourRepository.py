from Gruppe_18.src.main.repository.JSONRepository import JSONRepository

list_with_tours =

class TourRepository (JSONRepository):

    def filter_tour_by_location(tour_list, destination):
        filtered_tours = []
        for tour in tour_list:
            if tour['destination'] == destination:
                filtered_tours.append(tour)
        return filtered_tours

    def filter_tour_by_price(tour_list, max_price, min_price):
        filtered_tours = []
        for tour in tour_list:
            if min_price <= tour['cost'] <= max_price:
                filtered_tours.append(tour)
        return filtered_tours