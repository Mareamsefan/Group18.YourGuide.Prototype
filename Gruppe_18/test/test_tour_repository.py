import os
import uuid

from approvaltests.scrubbers import scrub_all_guids
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from Gruppe_18.src.main.model.models import db, Account

import pytest
from approvaltests import verify, Options

from Gruppe_18.src.main.repository.TourRepository import TourRepository
from Gruppe_18.src.main.repository.AccountRepository import AccountRepository
import datetime
from Gruppe_18.src.main.model.models import Tour



@pytest.fixture()
def get_session():
    engine = create_engine("sqlite:///Test.db", echo=True)

    Session = sessionmaker(bind=engine)

    return Session()

@pytest.fixture
def tour_re(get_session):
    return TourRepository(get_session)


@pytest.fixture()
def acc_rep(get_session):
    return AccountRepository(get_session)


@pytest.fixture()
def tour():
    return Tour(str(uuid.uuid4()), "Welcome to Dubai",
        datetime.date(2020, 10, 15),
        "Dubai",
        4,
        255,
        15,"English",
        "https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg")

# cancellation. return False
@pytest.fixture
def tour_2():
    return Tour(
        str(uuid.uuid4()),
    "Bergen Fjord Exploration",
    datetime.date(2024, 7, 10),
    "Bergen, Norway",
    4,
    4199,
    15,
    "English",
    "https://www.example.com/bergen-fjords.jpg"
)


@pytest.fixture
def tour_3():
    return Tour(
        str(uuid.uuid4()),
    "Discover Oslo's Charm",
    datetime.date(2024, 8, 5),
    "Oslo, Norway",
    2,
    2500,
    10,
    "Norwegian",
    "https://www.example.com/oslo-city.jpg"
)

@pytest.fixture()
def tour_4():
    return Tour(str(uuid.uuid4()),
    "Discover Oslo's Charm",
    datetime.date(2024, 8, 5),
    "Oslo, Norway",
    2,
    2500,
    0,
    "Norwegian",
    "https://www.example.com/oslo-city.jpg"
)

@pytest.fixture()
def guide():
    return Account("1", "guide", "guide", "guide", "12345678","guide@gmial.com")

@pytest.fixture()
def user():
    return Account("2", "user", "user", "user", "12345678", "user@gmial.com")

@pytest.fixture()
def admin():
    return Account("3", "admin", "admin", "admin", "12345678","admin@gmial.com")



@pytest.fixture()
def sqlalchemy_session(tour_re, acc_rep, tour, tour_2, tour_3, guide, user, admin):
    module_path = os.path.dirname(os.path.abspath(__file__))
    database_name = os.path.join(module_path, "Test.db")
    engine = create_engine(f"sqlite:///{database_name}", echo=True)

    session = sessionmaker(bind=engine)()

    db.metadata.create_all(bind=engine)
    tour_re.create_tour(tour)
    tour_re.create_tour(tour_2)
    tour_re.create_tour(tour_3)
    acc_rep.create_account(admin)
    acc_rep.create_account(user)
    acc_rep.create_account(guide)
    yield session

    session.close()
    db.metadata.drop_all(engine)


approval_options = Options().with_scrubber(scrub_all_guids)


# testing feature nonfunctional 1.19 and 1.20
def test_if_tour_is_created_saved_and_retrived(tour_re, sqlalchemy_session):
    saved_data = tour_re.get_all_tours()

    verify(saved_data, options=approval_options)


# testing feature nonfunctional 1.23.1
def test_if_booked_goes_up_by_one_after_registration_to_a_not_fully_booked_tour(tour_re, sqlalchemy_session):
    data = tour_re.get_all_tours()
    tour = data[0]
    tour_re.book_tour(tour)
    assert tour.booked == 1


# testing feature nonfunctional 1.24
def test_if_booking_to_fully_booked_tour_is_not_possible(tour_re, sqlalchemy_session, tour_4):
    tour_re.create_tour(tour_4)
    data = tour_re.get_all_tours()
    tour = data[len(data)-1]
    assert tour_re.book_tour(tour) == False


# testing feature nonfunctional 1.22
def test_if_booking_to_none_existing_tour_is_not_possible(tour_re, sqlalchemy_session, tour_4):
    assert tour_re.book_tour(tour_4) == False


# testing feature nonfunctional 1.23.2
def test_if_booked_goes_down_by_one_after_tour_cancelletion(tour_re, sqlalchemy_session):
    tours = tour_re.get_all_tours()
    tour = tours[0]
    tour_re.book_tour(tour)
    assert tour.booked == 1
    tour_re.cancel_booked_tour(tour)
    assert tour.booked == 0


# testing feature nonfunctional 1.17
def test_if_cancellation_on_none_existing_tour_is_not_possible(tour_re, sqlalchemy_session, tour_4):
    assert tour_re.cancel_booked_tour(tour_4) == False


# testing feature 1.19.1.3
def test_if_description_for_a_tour_is_correctly_returnet(tour_re, sqlalchemy_session):
    data = tour_re.get_all_tours()
    tour = data[0]
    assert tour_re.get_tour_description(tour.id) == "This tour will take you to Dubai for 4 " \
                                                    "hours, and is offered in English"


# testing feature nonfunctional 1.29
def test_if_getting_description_from_a_non_existing_tour_is_not_possible(tour_re, sqlalchemy_session):
    assert tour_re.get_tour_description("not_existing_id") == "Tour not found"


# testing feature nonfunctional 1.25
def test_if_tour_can_be_deleted_from_database(tour_re, sqlalchemy_session):
    data = tour_re.get_all_tours()
    assert len(data) == 3
    tour_re.delete_tour(data[0].id)
    saved_data = tour_re.get_all_tours()
    assert len(saved_data) == 2


# testing feature non-functional 1.27
def test_if_deleting_none_existing_tour_is_not_possible(tour_re, sqlalchemy_session):
    assert tour_re.delete_tour("not_existing_id") == False


# testing feature 1.8.5
def test_if_filtering_based_on_nothing_returns_all_tours(sqlalchemy_session, tour_re):
    filter_tour = tour_re.filter_combinations('', '', '', '')

    verify(filter_tour, options=approval_options)


# testing feature 1.8.3.1
def test_if_filtering_based_on_only_destination_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("Dubai", "", "", "")
    verify(filter_tour, options=approval_options)


# testing feature 1.8.3.2 and # testing feature 1.8.3.3
def test_if_filtering_based_on_only_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "500", "3000", "")
    verify(filter_tour, options=approval_options)


# testing feature 1.8.3.3
def test_if_filtering_only_on_max_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "", "3000", "")
    verify(filter_tour, options=approval_options)


# testing feature 1.8.3.2
def test_if_filtering_only_on_min_price_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "500", "", "")
    verify(filter_tour, options=approval_options)


# testing feature 1.8.3.4
def test_if_filtering_based_on_only_language_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("", "", "", "English")
    verify(filter_tour, options=approval_options)


# testing feature 1.8.4
def test_if_filtering_based_on_destination_price_and_language_is_as_expected(tour_re, sqlalchemy_session):
    filter_tour = tour_re.filter_combinations("Dubai", "0", "600", "English")
    verify(filter_tour, options=approval_options)


def test_if_getting_spesific_tour_is_possible(tour_re, sqlalchemy_session):
    all_tours = tour_re.get_all_tours()
    tour_1 = all_tours[0]
    tour_1_id = tour_1.id
    assert tour_1 == tour_re.get_specific_tour(tour_1_id)


# testing feature 1.8.5.1
def test_if_searching_tours_by_title_gives_is_as_expected(tour_re, sqlalchemy_session):
    searched_tour = tour_re.search_tour("dubai")
    verify(searched_tour, options=approval_options)


# testing feature nonfunctional 1.30.1
def test_if_guide_and_tour_relationship_gets_made_after_tour_creation(acc_rep, tour_re, sqlalchemy_session, guide):
    all_tours = tour_re.get_all_tours()
    tour_1 = all_tours[0]
    tour_1_id = tour_1.id
    user_id = guide.id
    assert tour_re.guide_register_to_tour(tour_1_id, user_id) == True


# testing feature nonfunctional 1.30.2
def test_if_guide_and_tour_relation_dosent_get_made_for_already_posted_tour(acc_rep, tour_re, sqlalchemy_session, guide):
    all_tours = tour_re.get_all_tours()
    tour_1 = all_tours[0]
    tour_1_id = tour_1.id
    user_id = guide.id
    tour_re.guide_register_to_tour(tour_1_id, user_id)

    all_tours = tour_re.get_all_tours()
    tour_1 = all_tours[0]
    tour_1_id = tour_1.id
    user_id = guide.id
    assert tour_re.guide_register_to_tour(tour_1_id, user_id) == None


# testing feature nonfunctional 1.30.3
def test_if_guide_tour_relation_dosent_get_made_if_tour_or_guide_dosent_exist(tour_re, sqlalchemy_session):
    assert tour_re.guide_register_to_tour("id", "id") == None


# testing feature nonfunctional 1.30.4
def test_if_guide_tour_relationship_gets_deleted_after_tour_get_deleted(tour_re, sqlalchemy_session, guide):
    all_tours = tour_re.get_all_tours()
    tour_1 = all_tours[0]
    tour_1_id = tour_1.id
    user_id = guide.id
    tour_re.guide_register_to_tour(tour_1_id, user_id)

    assert tour_re.guide_delete_tour(tour_1_id, user_id) == True


# testing feature nonfunctional 1.30.3
def test_if_deleting_relation_between_none_existing_tour_and_owner_is_not_possible(tour_re, sqlalchemy_session):
    assert tour_re.guide_delete_tour("id", "id") == None


# testing feature 1.18.10.1
def test_if_admin_gets_correct_number_of_total_users(tour_re, sqlalchemy_session):
    info = tour_re.admin_dashboard()
    assert info['num_users'] == 3


# testing feature 1.18.10.2
def test_if_admin_gets_correct_number_of_tours(tour_re, sqlalchemy_session):
    info = tour_re.admin_dashboard()
    assert info['num_tours'] == 3


# testing feature 1.18.10.4
def test_if_admin_gets_correct_number_of_guides(tour_re, sqlalchemy_session):
    info = tour_re.admin_dashboard()
    assert info['num_guides'] == 1


# testing feature 1.18.10.5
def test_if_admin_gets_correct_number_of_regular_users(tour_re, sqlalchemy_session):
    info = tour_re.admin_dashboard()
    assert info['num_regular_users'] == 1


# testing feature 1.18.10.6
def test_if_admin_gets_correct_number_of_admins(tour_re, sqlalchemy_session):
    info = tour_re.admin_dashboard()
    assert info['num_admin'] == 1


# testing feature 1.18.10.3
def test_if_admin_can_see_number_of_bookings(tour_re, sqlalchemy_session):
    tours = tour_re.get_all_tours()
    tour_re.book_tour(tours[0])
    info = tour_re.admin_dashboard()
    assert info['num_booked_tours'] == 1

