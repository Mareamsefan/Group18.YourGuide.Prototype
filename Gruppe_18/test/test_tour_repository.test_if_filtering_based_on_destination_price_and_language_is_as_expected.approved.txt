[(<guid_0>) Welcome to Dubai Dubai 4 255 English 15 https://www.hdwallpaper.nu/wp-content/uploads/2015/05/colosseum-1436103.jpg 0]
